### ISSUE TITLE
Proposal: method-level variance

### ISSUE THREAD

**Issue by masonwheeler** [dotnet/coreclr#24502]
It's the classic variance question, that I'm sure we've all seen a dozen times on StackOverflow: "why can't I pass a `List<Derived>` to a method parameter that takes `List<Base>` when `Derived` inherits from `Base`?"

Answering this formally would involve a bunch of complicated type theory, but the informal answer is simple and intuitive: "Because there's nothing stopping that method from calling `Add()` on that list with some other `Base`-derived object that's of a different class, and then you've violated type safety."

This is an issue that variance would really help with, but it's not available.  Variance only applies to delegates and interfaces, and only to the interface type itself in the latter case.  Various proposals and discussions have come about over the years regarding adding variance to class types, which tend to get shot down [with good reason](https://stackoverflow.com/questions/2733346/why-isnt-there-generic-variance-for-classes-in-c-sharp-4-0/2734070#2734070):

> Suppose you had a class `C<T>` that was covariant in T.  What might its implementation look like?  T has to be out only. That means that `C<T>` cannot have any method that takes a T, any property of type T with a setter, *or any field of type T*, because fields are logically the same as property setters; T goes in.  

The thing that seems strange to me is, why all this talk about variance on *class-level* type parameters, when the appropriate place to put it is clearly on *methods?*  Allow me to explain:

Going back to the fundamental variance question above, the problem is that our method might add an object of an inappropriate type to the `List<Base>` and violate type safety.  It would be safe to do so, though, if we had a guarantee that the method would not do that, which is exactly equivalent to contravariance.  Therefore, a method with the hypothetical signature  `void ReadBaseList(List<out Base> data)` be safe to pass a `List<Derived>` to.  Or, from the other direction, `void PopulateDerivedList(List<in Derived> list)` be safe to pass a `List<Base>` to, so long as the variance guarantee is maintained by the type system within the method body.

Implementing this would require two method-signature-level changes and four new type system rules, none of which are particularly onerous:

- Any method parameter with a generic type may optionally be declared as variant, as described above
- Any method with a generic type type parameter, either on itself or on its class, may have a variance constraint added to it
- It is an error for a method with a covariant constraint on itself to write to any of its own fields or properties whose type is generic, or to call any of its own methods which lack the covariant constraint.
- It is an error for a method with a contravariant constraint on itself to read from any of its own fields or properties whose type is generic, or to call any of its own methods which lack the contravariant constraint.
- The generic types `in T` and `out T` are not assignment-compatible with each other or with plain `T`. (ie. if you are inside a method with a parameter of type `List<in T>`, neither that list nor any of the objects inside of it can be passed to a method that takes a `T`, only to a method that takes a `in T`.)  But plain `T` can be assigned to both `in T` and `out T`.
- For the purposes of the previous rule, the presence of a variance constraint on a method converts its generic type parameter(s) into the variant equivalent.

Now we can finally resolve the question from the beginning.  Mark `List<T>.Add()` as contravariant, and then the method `void ReadBaseList(List<out Base> data)` is prevented from calling it because it has an incompatible `out Base` type, while `void PopulateDerivedList(List<in Derived> list)` can call it safely.  (But it would not be able to index into the list or retrieve an enumerator, which would be `out` operations.)

By moving the placement of variant type parameters on classes from the class level to the method level, we gain the ability to do interesting things with generics that have been unavailable up until now.  I believe this would be a useful addition to C# and to the CLR in general.

**Open questions:**

- How should variance constraints work when there are multiple type parameters?
- Is there a place for class-level variance constraints, perhaps for immutable types?
- What would the syntax for a variance constraint look like in C#?

**Comment by YairHalberstadt** [comment]
Why not just declare your method as accepting an IReadOnlyList<Base>?

**Comment by masonwheeler** [comment]
For this specific example, sure, that might work.  For the general principle that `List<T>` is being used as a simple, intuitive illustration of, however...

**Comment by svick** [comment]
I'm struggling to understand what would this proposal enable that's not possible today. What are those "interesting things" you mention?

For example, couldn't `ReadBaseList` be defined as `void ReadBaseList<T>(List<T> data) where T : Base`?

I don't know how to use the same approach for `PopulateDerivedList`, because there is no constraint for "`T` has to be the base type of `Derived`". But I think adding that constraint would be a much simpler extension of C# than this proposal.

**Comment by YairHalberstadt** [comment]
Assuming that some form of shapes/type classes/ extension interfaces was defined, wouldn't it be possible in all cases to define a covariant interface with only the methods you plan to use, and then use a shape to witness the interface on the type.

If this is particularly onerous, a language level feature might be able to do this all for you, without requiring runtime changes.

Would this be an acceptable alternative?

**Comment by masonwheeler** [comment]
> Would this be an acceptable alternative?

No, for the same reason I don't think shapes or other changes to the type system should be implemented inside the C# compiler: because then you'd have interoperability problems with other CLR languages.

**Comment by masonwheeler** [comment]
@svick 

>For example, couldn't `ReadBaseList` be defined as `void ReadBaseList<T>(List<T> data) where T : Base`?

No, this would have two problems:

1) It doesn't provide a guarantee that nothing will be added
2) It takes what used to be a non-generic method and turns it into a generic method, adding complexity to the codegen

> I don't know how to use the same approach for `PopulateDerivedList`, because there is no constraint for "`T` has to be the base type of `Derived`". But I think adding that constraint would be a much simpler extension of C# than this proposal.

I'm not sure about that.  It seems to me that adding that constraint would pretty much be exactly equivalent to (the contravariant half of) this proposal in terms of complexity.

**Comment by svick** [comment]
@masonwheeler 

> No, for the same reason I don't think shapes or other changes to the type system should be implemented inside the C# compiler: because then you'd have interoperability problems with other CLR languages.

How is this proposal any better than that? It's also a change to the type system, one that would probably require CLR support and would have interoperability problems with other languages. If you think it wouldn't, then I think you should explain why and how exactly would this proposal compile using the existing type system.

**Comment by masonwheeler** [comment]
@svick I think you misunderstand:  I'm saying that as a general principle, changes to the type system should be implemented *in the CLR type system itself,* rather than simply hacked into the C# compiler, so as to give all the languages the same baseline to build off of.

**Comment by svick** [comment]
@masonwheeler You're right, I misunderstood (I missed that this is coreclr, not csharplang).

Does that mean your opinion on type classes would change if they were a CLR feature? How would they then compare with this proposal?

**Comment by masonwheeler** [comment]
> Does that mean your opinion on type classes would change if they were a CLR feature? 

What change?  [That's been my position all along.](https://github.com/dotnet/csharplang/issues/339#issuecomment-400112464)
